src/crush/balance.c

```diff
@@ -0,0 +1,195 @@
+#include <assert.h>
+#include <float.h>
+#include <memory.h>
+
+#include "crush_compat.h"
+#include "int_types.h"
+
+#include "balance.h"
+
+#ifndef MAX
+#define MAX(a, b) ((a) > (b) ? (a) : (b))
+#endif // MAX
+
+#ifndef MIN
+#define MIN(a, b) ((a) < (b) ? (a) : (b))
+#endif // MIN
+
+int balance_values(int values_count, int items_count, __s64* original_straws, __u32* target_weights, __u32* weights)
+{
+  __s64 straws[values_count * items_count];
+  for (int i = 0; i < values_count; i++)
+    for (int j = 0; j < items_count; j++)
+      straws[i * items_count + j] = div64_s64(original_straws[i * items_count + j], weights[j]);
+  // a single item is always perfectly balanced
+  if (items_count < 2)
+    return 0;
+
+  // how many values are expected for each item ?
+  __u64 total_weight = 0;
+  for (int i = 0; i < items_count; i++)
+    total_weight += target_weights[i];
+  int expected[items_count];
+  int total_expected = 0;
+  for (int i = 0; i < items_count; i++) {
+    double normalized_weight = (double)target_weights[i] / total_weight;
+    expected[i] = values_count * normalized_weight;
+    total_expected += expected[i];
+  }
+  assert(values_count - total_expected < items_count);
+  for (int i = 0; i < values_count - total_expected; i++)
+    expected[i]++;
+
+  int max_iterations = CRUSH_BALANCE_MAX_ITERATIONS + values_count;
+  int iterations;
+  for (iterations = 0; iterations < max_iterations; iterations++) {
+    int delta[items_count];
+    memcpy(delta, expected, sizeof(int) * items_count);
+
+    // One of the values that landed on each item because it
+    // got a straw that is not much higher than the second best straw.
+    // This is the closest winner. And the closest looser is the item
```

Comment:
since we are talking about winners and losers here, we could consider s/looser/loser/g here, (nitpick, feel free to ignore)

Reply:
fixed thanks :-)